feat: add FileBackedTaskManager for saving Tasks whith file

alekseevwork opened the pull request.

alekseevwork committed:
feat: add FileBackedTaskManager for saving Tasks whith file

alekseevwork committed:
fix: fixed code style

Alexander-Bolotov commented on src/taskmanager/exception/ManagerSaveException.java:
```diff
@@ -0,0 +1,11 @@
+package taskmanager.exception;
```
Пекеджи лучше называть в одном стиле. Во множественном числе, как и остальные (или наоборот)

Alexander-Bolotov commented on src/taskmanager/managers/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,163 @@
+import taskmanager.tasks.Subtask;
+import taskmanager.tasks.Task;
+
+import java.io.*;
```
Лучше не использовать rawType при импортах. Что бы идея не делала это автоматически нужно изменить в ней настройку

Alexander-Bolotov commented on src/taskmanager/managers/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,163 @@
+            System.out.println(e.getMessage());
+            throw new ManagerSaveException(e.getMessage(), (IOException) e.getCause());
+        } finally {
+            save();
```
Почему поместил в блок finally? Обычно он используется для закрытия ресурсов, но у тебя тут try-with-resources поэтому этого не требуется.
Лучше перенести save в try

Alexander-Bolotov commented on test/taskmanager/managers/FileBackedTaskManagerTest.java:
```diff
@@ -0,0 +1,106 @@
+
+class FileBackedTaskManagerTest {
+
+    private static Map<Integer, Task> tasks;
+    private static Map<Integer, Epic> epics;
+    private static Map<Integer, Subtask> subtasks;
```
это не должно быть полем теста

Alexander-Bolotov commented on test/taskmanager/managers/FileBackedTaskManagerTest.java:
```diff
@@ -0,0 +1,106 @@
+    private static Map<Integer, Epic> epics;
+    private static Map<Integer, Subtask> subtasks;
+
+    static TaskManager manager;
```
а это ок)

Alexander-Bolotov commented on test/taskmanager/managers/FileBackedTaskManagerTest.java:
```diff
@@ -0,0 +1,106 @@
+
+    static TaskManager manager;
+
+    static Task task;
+    static Epic epic;
+    static Subtask subtask;
```
а это опять не должно)

alekseevwork committed:
delete static fields within the FileBackedTaskManagerTest class,
remove the finally structure in the loadFromFile method of the FileBackedTaskManager class,
renamed package exception -> exceptions.
Signed-off-by: Artem <alekseevwork7@gmail.com>

avfyodorov commented on src/taskmanager/exceptions/ManagerSaveException.java:
```diff
@@ -0,0 +1,11 @@
+import java.io.IOException;
+import java.io.UncheckedIOException;
+
+public class ManagerSaveException extends UncheckedIOException {
```
UncheckedIOException 

В принципе, можно было бы наследоваться от обычного runtime Exception. Этого вполне достаточно для проброса непроверяемого исключения. 
Необязательно, на Ваше усмотрение.

alekseevwork replied:
Сделано)

avfyodorov commented on src/taskmanager/managers/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,166 @@
+import java.nio.file.Paths;
+import java.util.Map;
+
+public class FileBackedTaskManager extends InMemoryTaskManager implements TaskManager {
```
implements TaskManager {

Реализовывать интерфейс не нужно. Он уже реализован в классе-предке.
Здесь достаточно будет переписать только те методы, у которых изменилось поведение.

alekseevwork replied:
Понял, сделал

avfyodorov commented on src/taskmanager/managers/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,166 @@
+
+public class FileBackedTaskManager extends InMemoryTaskManager implements TaskManager {
+
+    private static Path taskFile = Paths.get("tasksFile.csv");
```
Paths.get("tasksFile.csv");

Лучше было бы перенести этот файл из корня проекта в отдельный подкаталог с ресурсами, например, resources/data.csv.
Держать исходный код вместе с рабочими данными - это плохая практика.

avfyodorov replied:
static Path taskFile

Не надо делать эту переменную static. Если запустить несколько экземпляров менеджера, то данные будут записываться в один и тот же файл и всё запутается. )

avfyodorov commented on src/taskmanager/managers/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,166 @@
+            taskFile = Files.createFile(Paths.get(taskFile.toUri()));
+        }
+        try {
+            loadFromFile(taskFile);
```
loadFromFile(taskFile);

Лучше не запускать внутри конструктора загрузку данных из файлов. Потому что если что-то пойдёт не так, объект окажется недосознанным и придётся долго разбираться, что к чему и откуда взялось. Здесь можно ограничиться: 

	private final File file;

	public FileBackedTaskManager(File file) {
		this.file = file;
	}

avfyodorov commented on src/taskmanager/managers/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,166 @@
+            }
+            save();
+        } catch (IOException e) {
+            System.out.println(e.getMessage());
```
Выводить на экран здесь ничего не нужно, достаточно пробросить исключение.

avfyodorov commented on src/taskmanager/managers/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,166 @@
+        }
+    }
+
+    public static void loadFromFile(Path file) throws IOException {
```
throws IOException

Это не нужно, потому что данное исключение перехватывается внутри метода, а пробрасывается самодельное.

avfyodorov commented on src/taskmanager/managers/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,166 @@
+        }
+    }
+
+    static void save() throws ManagerSaveException {
```
throws ManagerSaveException

Здесь и дальше в аналогичных местах программы. 
Это лишняя конструкция, потому что самодельное исключение -  непроверяемое, оно само пробросится.

avfyodorov commented on src/taskmanager/managers/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,166 @@
+    }
+
+    @Override
+    public Task getTasksForId(int taskId) {
```
Здесь и дальше. 

Однострочные методы, состоящие из одной super, не добавляют никакой новой функциональности. Поэтому они не нужны. От них можно спокойно избавиться.

avfyodorov reviewed (commented):
Добрый день, Артём! 

В целом работа хорошая, правда, есть несколько небольших уточнений. 

Есть просьба-оставлять хоть какую-то реакцию на мои комментарии. Сделано/ Не сделано, потому что не согласен/ Не заметил. 🙂

alekseevwork committed:
fix: Change the loadFromFile method in the FileBackedTaskManager class and remove unnecessary methods.

avfyodorov reviewed (commented):
Добрый день, Артём!

Замечаний нет.
Работа принята.
